xlators/mgmt/glusterd/src/glusterd-utils.c

```diff
@@ -5265,9 +5266,12 @@ glusterd_import_friend_volumes_synctask(void *opaque)
     conf->restart_bricks = _gf_true;
 
     while (i <= count) {
-        ret = glusterd_import_friend_volume(peer_data, i);
-        if (ret) {
-            break;
+        /*Check if volume has any peer update */
+        update = (1UL & (arg->status_arr[(i / 64)] >> (i % 64)));
+        if (update) {
+            ret = glusterd_import_friend_volume(peer_data, i, arg);
+            if (ret)
+                break;
         }
         i++;
     }
```

Comment:
This is not what I had in mind when I said we should avoid checking all volumes. This way you are actually checking every single volume.

I would use something like this:

```c
while (i <= count) {
    bm = arg->status_arr[i / 64];
    while (bm != 0) {
        mask = bm & (-bm); /* mask will contain the lowest bit set from bm. */
        bm ^= mask;
        ret = glusterd_import_friend_volume(peer_data, i + ffsll(mask), arg);
        if (ret < 0) {
            break;
        }
    }
    i += 64;
}
```

This way we skip all non-modified volume very fast. It's at least 64 times faster if we don't have any change.